compiler/rustc_codegen_ssa/src/back/archive.rs

```diff
@@ -1,35 +1,39 @@
+use rustc_data_structures::memmap::Mmap;
 use rustc_session::cstore::DllImport;
 use rustc_session::Session;
 
 use object::read::archive::ArchiveFile;
-use rustc_data_structures::memmap::Mmap;
+
+use std::fmt::Display;
 use std::fs::File;
 use std::io;
 use std::path::{Path, PathBuf};
 
+pub fn get_possible_native_lib_names(verbatim: bool, name: &str, sess: &Session) -> Vec<String> {
+    // On Windows, static libraries sometimes show up as libfoo.a and other
+    // times show up as foo.lib
+    vec![
+        if verbatim {
```

Comment:
In case of `verbatim` `lib{}.a` should be skipped, right? Also would it be possible to store the filename in the crate metadata instead of attempting to predict it?

Reply:
>In case of verbatim lib{}.a should be skipped, right?

If we are *not* using metadata, then we'd ideally redo the same logic as in `fn find_library`, but inside the rlib archive:
- try to extract member named `verbatim ? "{name}" : "{prefix}{name}{suffix}"`
- if there's no such member then extract `"lib{name}.a"`

Reply:
>Also would it be possible to store the filename in the crate metadata instead of attempting to predict it?

If we *are* using metadata, then we'd just record the name returned by `find_library` to metadata, I'm just not exactly sure whether we can do that in presence of things like pipelined compilation.
The metadata is formed and written long before the work of link.rs, are you sure all the static libraries to find with `find_library` necessarily exist at at that point in time?

Reply:
> The metadata is formed and written long before the work of link.rs, are you sure all the static libraries to find with find_library necessarily exist at at that point in time?

Pretty sure. These static libraries are only native static libraries, which don't participate in pipelined compilation.

Reply:
Native static libraries are often produced by cargo build scripts.
If we want to use `find_library` from metadata, then we are saying that all build script results (from all dependent crates?) must be ready at that point.
Yeah, it seems like that is already true.

Reply:
> Native static libraries are often produced by cargo build scripts.

Build scripts have already run by the time rustc is invoked. This is necessary as build scripts can influence which arguments are passed to rustc. Also there is no way for cargo to signal rustc that artifacts are ready, so anything the current rustc invocation may ever need whether before the linking step or during must already be compiled by the time rustc is invoked. Cargo makes sure to not pipeline anything that will invoke the system linker for this reason.